src/mpid/ch4/shm/ipc/src/ipc_win.c

```diff
@@ -149,14 +157,27 @@ int MPIDI_IPC_mpi_win_create_hook(MPIR_Win * win)
         if (i == shm_comm_ptr->rank) {
             shared_table[i].shm_base_addr = win->base;
         } else {
-            /* Attach only remote buffer. No-op if ipc_type is NONE including zero-size. */
-            int node_rank = ranks_in_shm_grp[shm_comm_ptr->local_size + i];
-            mpi_errno = MPIDI_IPCI_attach_mem(ipc_shared_table[i].ipc_type, node_rank,
-                                              ipc_shared_table[i].mem_handle,
-                                              attr.gpu_attr.device,
-                                              ipc_shared_table[i].size,
-                                              &shared_table[i].shm_base_addr);
-            MPIR_ERR_CHECK(mpi_errno);
+            /* attach remote buffer */
+            switch (ipc_shared_table[i].ipc_type) {
+                case MPIDI_IPCI_TYPE__XPMEM:
+                    mpi_errno =
+                        MPIDI_XPMEM_ipc_handle_map(ipc_shared_table[i].ipc_handle.xpmem,
+                                                   &shared_table[i].shm_base_addr);
+                    break;
+                case MPIDI_IPCI_TYPE__GPU:
+                    mpi_errno =
+                        MPIDI_GPU_ipc_handle_map(ipc_shared_table[i].ipc_handle.gpu,
+                                                 ipc_attr.gpu_attr.device, MPI_INT,
+                                                 &shared_table[i].shm_base_addr);
```

Comment:
This code doesn't seem correct.  You are mapping the window against the device on which the target buffer resides.  We need to map each remote process buffer against each of our local GPUs.

Reply:
This code actually implements what you mean. `ipc_attr.gpu_attr.device` is the local GPU device where `win->base` resides, and `ipc_shared_table[i].ipc_handle.gpu` stands for each remote process window buffer.

Reply:
I think you are misunderstanding.  Each remote process window buffer needs to be mapped to each local device (not just the local GPU device where `win->base` resides.

Reply:
Got it, I can add a comment there for future fixup